### PR TITLE
Allow sending of browser logs to devtools.

### PR DESCRIPTION
**Pull request by jellefoks**
This adds a debug_log() console command that can turn on sending
browser logs to attached debuggers.
    
This also adds related working buttons on the Cobalt page in devtools.
    
b/251825168

### CODE REVIEW

File: cobalt/browser/debug_console.cc
Comment (andrewsavage1) [inline cobalt/browser/debug_console.cc:210]: Slight pref for setting an int in the conditional (i.e. num_debuggers) and doing modulo that variable outside of the conditional

File: cobalt/browser/web_module.cc
Comment (andrewsavage1) [inline cobalt/browser/web_module.cc:1145]: I think we always use braces: go/cstyle#Formatting_Looping_Branching

File: cobalt/debug/remote/debug_web_server.cc
Comment (andrewsavage1) [inline cobalt/debug/remote/debug_web_server.cc:122]: Do you want to keep these all as regular logs?

**Review by andrewsavage1 — COMMENTED** [review]
Will this add the browser logs directly to the console or somewhere else? Also, it would be nice if some of the changes to already large functions were factored out into their own functions, but I'll leave that up to you

File: cobalt/browser/web_module.cc
Comment (jellefoks) [reply cobalt/browser/web_module.cc:1145]: This matches the exception in the cstyle because it fits on one line.

File: cobalt/debug/remote/debug_web_server.cc
Comment (jellefoks) [reply cobalt/debug/remote/debug_web_server.cc:122]: No, this should be removed. I'll update the CL once the parent merges.

File: cobalt/browser/debug_console.cc
Comment (jellefoks) [reply cobalt/browser/debug_console.cc:210]: Agreed. I'll update the CL once the parent merges.

**Comment by jellefoks** [comment]
> Will this add the browser logs directly to the console or somewhere else? Also, it would be nice if some of the changes to already large functions were factored out into their own functions, but I'll leave that up to you

This will allow, as the description says, sending of the browser logs to attached debuggers. devtools debuggers to be precise. With enable/disable buttons.

File: cobalt/debug/remote/debug_web_server.cc
Comment (jellefoks) [inline cobalt/debug/remote/debug_web_server.cc:134]: This too should probably be kept a DLOG, it's too spammy.

File: cobalt/browser/debug_console.cc
Comment (jellefoks) [reply cobalt/browser/debug_console.cc:210]: Oh, actually it's not possible, the enum is generated from an IDL.

cobalt/debug/console/debug_console_mode.idl

Comment (jellefoks) [reply cobalt/browser/debug_console.cc:210]: Updated as suggested in the parent

File: cobalt/debug/backend/debug_module.cc
Comment (sherryzy) [inline cobalt/debug/backend/debug_module.cc:1]: Thanks for doing all these code improvement. But the same as the comments below, it will be great if we can keep the pure code refactor in a separate PR.

File: cobalt/debug/console/command_manager.cc
Comment (jellefoks) [reply cobalt/debug/console/command_manager.cc:50]: While not spelled out as such in a style guide, it's a common pattern in Cobalt and Chrome to document static methods with this comment right above the implementation.

See Chrome: https://source.chromium.org/search?q=f:%5C.cc%20%5E%2F%2F%5Cs%2Bstatic%20-f:%5Eout%2F&ss=chromium

And it's actively maintained, see for example https://chromium-review.googlesource.com/c/chromium/src/+/2830312

File: cobalt/debug/backend/debug_module.cc
Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: Having to create and juggle separate PRs for code cleanup will result in cleanups not happening.

File: cobalt/web/window_timers_test.cc
Comment (jellefoks) [reply cobalt/web/window_timers_test.cc:55]: Having to create and juggle separate PRs for code cleanup will result in cleanups not happening.

File: cobalt/debug/console/command_manager.cc
Comment (jellefoks) [reply cobalt/debug/console/command_manager.cc:58]: This passed clang format and linter and that should be enough for formatting requirements.

**Review by andrewsavage1 — APPROVED** [review]

File: cobalt/debug/backend/debug_module.cc
Comment (sherryzy) [reply cobalt/debug/backend/debug_module.cc:1]: Well, I'm not totally agree with that. It should be as simple as creating a code cleanup PR first and then the new PR. For the pure cleanup PR should get an approve very fast. 
Suppose the new change cause any issue, and we need to revert that change, then separating the cleanup part can ensure the cleanup part still get in.

File: cobalt/debug/console/command_manager.cc
Comment (sherryzy) [reply cobalt/debug/console/command_manager.cc:50]: Well, it says prefer not to but keep consistent with surrounding code: go/c-readability-advice#static-comments. Good to know we choose to do this way in Cobalt and Chrome....

File: cobalt/debug/backend/debug_module.cc
Comment (sherryzy) [reply cobalt/debug/backend/debug_module.cc:1]: Also if "waiting for the tests, dealing with re-running the flakes, etc" is preventing send out a code change, I feel we probably should fix this.....

Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: I looked at all merged commits this year, and outside of fix-it week I found only *one* pure cleanup commit, and it removed two lines of comments from a python script. If it actually was a low threshold, I would think that we'd see more of them.

And yes, I wish that tests ran fast and didn't have flakes.

Comment (kaidokert) [reply cobalt/debug/backend/debug_module.cc:1]: @jellefoks i'm going to chime in here - please isolate clean-ups into separate PRs. There are plenty of clean-ups going in at a pretty rapid clip right now, and we have previous team retrospective outcomes where people have been asking to do just that b/230661351

I totally do get the point that this used to be hard on our internal CI, but in the new world things are far more traceable and quicker, and if a cleanup breaks something ( like mine did just yesterday https://github.com/youtube/cobalt/pull/1057 ) it's quick and cheap to detect and revert. 

Please isolate clean-ups, and i'd be happy to see a lot more of those going into trunk.

Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: My point is nobody does, and in practice that means people don't do cleanups.

File: cobalt/debug/console/command_manager.cc
Comment (jellefoks) [reply cobalt/debug/console/command_manager.cc:50]: Thanks for the link.

File: cobalt/debug/backend/debug_module.cc
Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: Very well, cleanups removed.

Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: To my point, after removing the cleanups, it re-runs the tests and hits two flakes totally not related to this CL... 

1) player_filter_tests fails in the `AssistantManage` platform dependent thread

https://github.com/youtube/cobalt/actions/runs/5695693683/job/15440095280?pr=656

2)  adb: device '8D084B07' not found

https://github.com/youtube/cobalt/actions/runs/5695693683/job/15440095280?pr=656

Comment (sherryzy) [reply cobalt/debug/backend/debug_module.cc:1]: Oh no, don't just remove those cleanups, but instead please send it out just in a separate PR. Do code improvement / cleanup should really be encouraged as part of better engineering work and it's great work!

> My point is nobody does, and in practice that means people don't do cleanups.

My feeling for this is that:
1. We probably should get the better engineering work more recognized, like the cleaning up work you did in this PR.
2. The flaky test issues. That's mostly those on device tests right? If it's just some simple code clean up or refactor work, do we need to run those tests? (I actually have no idea when we need to run on device tests...)

Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: Rebased to re-run the tests, and two new  flakes totally not related to this CL...

1)  linux-x64 / x64_wpt_test  Failed in various WebPlatformTest.Run/websockets_*

https://github.com/youtube/cobalt/actions/runs/5696267018/job/15441253175?pr=656
  
2)  SbSocketResolveTest. IgnoreExtraBits on evergreen-arm-hardfp / sbversion-14_on_device_2
 
https://github.com/youtube/cobalt/actions/runs/5696267023/job/15441340604?pr=656

Addressing all of those goes well beyond the scope of this change, or a cleanup CL and is not something that is simple, fast, and easy to deal with just to clean up some code.

Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: Round three. This round's flakes totally not related to this CL...

1)  SbAudioSinkTest. ContinuousAppend on android-arm64 / arm64_on_device_2

https://github.com/youtube/cobalt/actions/runs/5696828082/job/15443037982?pr=656

2) `Invalid configuration for device 19171HFDD4335J` on android-arm / arm_on_device_3

https://github.com/youtube/cobalt/actions/runs/5696828082/job/15442895806?pr=656

(runs not finished yet, will update on monday if others fail).

I think this should make it clear that it's flaky, slow, and high effort. Every flake it different, requires opening multiple pages to find where the issue is logged, and often involves scanning long log files. I can't afford to spend that much time on just a cleanup PR. If it didn't go in with another change, it's not going in.

Comment (andrewsavage1) [reply cobalt/debug/backend/debug_module.cc:1]: We haven't required on device tests because it is so flaky—I don't know the general consensus, but if the odt failure looks unrelated to what I'm doing, I'll ignore them and submit my code. Also fwiw you can re-run individual flaky jobs using the GItHub UI instead of rebasing and kicking off all of them again, that way you can more easily double check that something is a flake or a real failure without running into a different issue

Comment (jellefoks) [reply cobalt/debug/backend/debug_module.cc:1]: I agree, but it's also kind of sad that they are flaky and slow. I found odt's would be important for this change because of the web debugger tests. Similarly would be the code cleanup that I removed from this PR be best covered by the odt's.

**Pull request merged**
